lib/Infer/EnumerativeSynthesis.cpp

```diff
@@ -114,12 +124,25 @@ namespace {
 // experiment with synthesizing at reduced bitwidth, then expanding the result
 // aggressively avoid calling into the solver
 
-void addGuess(Inst *RHS, int MaxCost, std::vector<Inst *> &Guesses,
+void addGuess(Inst *RHS, unsigned TargetWidth, InstContext &IC, int MaxCost, std::vector<Inst *> &Guesses,
               int &TooExpensive) {
-  if (souper::cost(RHS) < MaxCost)
-    Guesses.push_back(RHS);
-  else
-    TooExpensive++;
+  if (TargetWidth > RHS->Width) {
+    auto NSExt = IC.getInst(Inst::SExt, TargetWidth, { RHS });
+    auto NZExt = IC.getInst(Inst::ZExt, TargetWidth, { RHS });
+    addGuess(NSExt, TargetWidth, IC, MaxCost, Guesses, TooExpensive);
```

Comment:
why do you have a recursive addguess call before incrementing the cost? this seems wrong

Reply:
what do you mean "before incrementing the cost"? These recursive calls are a way to convert an existing about-to-be-added guess which doesn't matches the `TargetWidth`. The recursive call will make sure that the new converted guess is still less in cost than the LHS

Reply:
ok good